File: src/native/androidProject/app/src/main/java/io/yourname/androidproject/NativeCameraManager.kt
package io.yourname.androidproject

import android.graphics.Color
import android.graphics.RectF
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.MotionEvent
import android.view.View
import android.webkit.WebView
import android.widget.TextView
import androidx.camera.view.PreviewView
import com.google.mlkit.vision.barcode.common.Barcode
import io.yourname.androidproject.camera.*
import io.yourname.androidproject.utils.BridgeUtils
import org.json.JSONObject

/**
 * Thin facade — wires all camera sub-components and exposes the public API
 * that NativeBridge.kt calls. All logic lives in the camera/ package.
 */
class NativeCameraManager(
    private val activity: MainActivity,
    private val previewView: PreviewView,
    private val webView: WebView,
    private val debugOverlay: View? = null,
    private val debugQrStatus: TextView? = null,
    private val debugBarcodeOverlay: View? = null
) {

    companion object {
        private const val TAG = "NativeCameraManager"
        const val PERMISSION_REQUEST_CODE = CameraSessionManager.PERMISSION_REQUEST_CODE

        // QR overlay colors — change here to update both states
        private const val COLOR_QR_IN_FRAME = "#5500FF00"  // green, semi-transparent
        private const val COLOR_QR_DETECTED  = "#55FF6600"  // orange, semi-transparent
        private const val OVERLAY_HIDE_DELAY_MS = 200L
    }

    private val stateMachine = VideoStreamStateMachine()

    private val zoomController = ZoomController(
        context = activity,
        stateMachine = stateMachine,
        onZoomChanged = { zoomLevel, minZoom, maxZoom ->
            val payload = JSONObject().apply {
                // zoomLevel = raw CameraX zoomRatio (1.0=1x, 2.0=2x)
                put("zoomLevel", Math.round(zoomLevel * 10.0) / 10.0)
                put("minZoom", Math.round(minZoom * 10.0) / 10.0)
                put("maxZoom", Math.round(maxZoom * 10.0) / 10.0)
            }
            BridgeUtils.notifyWebJson(webView, BridgeUtils.WebEvents.ON_ZOOM_CHANGED, payload)
            // Repaint overlay at new zoom level — barcode rect shifts as camera zooms
            if (showQrDetected && overlayVisible) {
                lastBarcodeScreenRect?.let { repaintQrOverlay(it, overlayIsOrange) }
            }
        }
    )

    private val torchController = TorchController(
        onTorchChanged = { enabled ->
            val payload = JSONObject().apply { put("enabled", enabled) }
            BridgeUtils.notifyWebJson(webView, BridgeUtils.WebEvents.ON_TORCH_CHANGED, payload)
        }
    )

    private val barcodeDetector = BarcodeDetector(
        zoomController = zoomController,
        onDetected = { barcode, imgW, imgH -> onBarcodeDetected(barcode, imgW, imgH) }
    )

    private val sessionManager = CameraSessionManager(
        activity = activity,
        previewView = previewView,
        stateMachine = stateMachine,
        zoomController = zoomController,
        torchController = torchController,
        barcodeDetector = barcodeDetector,
        onReady = {
            BridgeUtils.notifyWeb(webView, BridgeUtils.WebEvents.ON_VIDEO_STREAM_READY)
        },
        onStopped = {
            debugOverlay?.let { activity.runOnUiThread { it.visibility = View.GONE } }
            BridgeUtils.notifyWeb(webView, BridgeUtils.WebEvents.ON_VIDEO_STREAM_STOPPED)
        },
        onError = { message ->
            BridgeUtils.notifyWebError(webView, BridgeUtils.WebEvents.ON_CAMERA_ERROR, message)
        }
    )

    private val holdController = HoldController(
        stateMachine = stateMachine,
        barcodeDetector = barcodeDetector
    )

    // Viewfinder state
    private var viewfinderScreenRect: RectF? = null

    // QR detected overlay state
    private var showQrDetected: Boolean = false
    private var overlayIsOrange: Boolean = false       // logical state — set before showQrOverlay
    private var overlayPaintedOrange: Boolean = false  // what color was last painted
    private var overlayVisible: Boolean = false
    private var lastBarcodeScreenRect: RectF? = null
    private val overlayHideHandler = Handler(Looper.getMainLooper())
    private var overlayHideRunnable: Runnable? = null

    init {
        stateMachine.addListener(object : VideoStreamStateListener {
            override fun onStateChanged(prev: VideoStreamState, next: VideoStreamState) {
                // Reset overlay when hold ends (new scan cycle) or stream stops
                if ((prev == VideoStreamState.HOLD && next == VideoStreamState.STREAMING) ||
                    next == VideoStreamState.IDLE || next == VideoStreamState.STOPPING
                ) {
                    hideQrOverlay()
                }
            }
        })
    }

    // ---- Public API (called from NativeBridge) ----

    fun start(facing: String = "back", viewfinderRectJson: JSONObject? = null, zoomOptions: JSONObject? = null, scanFormat: String = "all", fpsMin: Int? = null, fpsMax: Int? = null, showQrDetected: Boolean = false) {
        if (stateMachine.isActive) return

        this.showQrDetected = showQrDetected
        this.overlayIsOrange = false
        this.overlayPaintedOrange = false
        this.overlayVisible = false
        this.lastBarcodeScreenRect = null

        val autoZoom = zoomOptions?.optBoolean("auto", false) ?: false
        val initialZoomPct = zoomOptions?.optDouble("initial", 1.0)?.toFloat() ?: 1.0f

        // Parse viewfinder rect and position debug overlay
        viewfinderRectJson?.let { json ->
            val result = ViewfinderMapper.parseViewfinderRect(json, webView)
            viewfinderScreenRect = result?.first
            result?.second?.let { jsRect -> positionDebugOverlay(jsRect) }
        } ?: run {
            viewfinderScreenRect = null
            Log.w(TAG, "No viewfinderRect — all QR codes will be reported")
        }

        stateMachine.transition(VideoStreamState.STARTING)
        sessionManager.start(
            facing = facing,
            autoZoom = autoZoom,
            initialZoomPct = initialZoomPct,  // now Float multiplier (1.0 = 1x)
            scanFormat = scanFormat,
            fpsMin = fpsMin,
            fpsMax = fpsMax
        )
    }

    fun stop() {
        if (!stateMachine.isActive) return
        holdController.reset()
        stateMachine.transition(VideoStreamState.STOPPING)
        sessionManager.stop()
    }

    fun flip() {
        if (!stateMachine.isActive) return
        holdController.reset()
        stateMachine.transition(VideoStreamState.FLIPPING)
        sessionManager.flip()
    }

    fun setZoom(multiplier: Float) = zoomController.setZoom(multiplier)

    fun setTorch(on: Boolean) = torchController.setTorch(on)

    fun setFps(min: Int?, max: Int?) {
        if (!stateMachine.isActive) return
        sessionManager.setFps(min, max)
    }

    fun onTouchEvent(event: MotionEvent): Boolean = zoomController.onTouchEvent(event)

    fun onPermissionResult(granted: Boolean) = sessionManager.onPermissionResult(granted)

    fun cleanup() = sessionManager.cleanup()

    // ---- Detection handler ----

    private fun onBarcodeDetected(barcode: Barcode, imageWidth: Int, imageHeight: Int) {
        val value = barcode.rawValue ?: return

        // Hold gate — defensive check for in-flight frames
        if (stateMachine.state == VideoStreamState.HOLD) {
            Log.d(TAG, "onBarcodeDetected — in hold, skipping: $value")
            return
        }

        // QR overlay — show green immediately (QR in frame), update position each frame
        val barcodeScreenRect = barcode.boundingBox?.let { box ->
            ViewfinderMapper.mapBarcodeToScreen(
                box.left, box.top, box.right, box.bottom,
                imageWidth, imageHeight,
                previewView.width.toFloat(), previewView.height.toFloat()
            )
        }
        if (showQrDetected && barcodeScreenRect != null) {
            showQrOverlay(barcodeScreenRect, orange = overlayIsOrange)
        }

        // Viewfinder region filter
        val regionRect = viewfinderScreenRect
        if (regionRect != null) {
            if (barcodeScreenRect != null) {
                if (!regionRect.contains(barcodeScreenRect)) {
                    updateDebugQrStatus(inside = false)
                    return
                }
                updateDebugQrStatus(inside = true)
            }
        }

        // Same-value dedup — enter hold either way
        if (!holdController.isNewValue(value)) {
            Log.d(TAG, "onBarcodeDetected — same value, suppressing: $value")
            holdController.startHold()
            return
        }

        // New value confirmed — flip overlay to orange before firing event
        if (showQrDetected && barcodeScreenRect != null) {
            overlayIsOrange = true
            showQrOverlay(barcodeScreenRect, orange = true)
        }

        holdController.startHold()

        val payload = JSONObject().apply {
            put("value", value)
            put("format", BarcodeDetector.formatName(barcode.format))
        }
        BridgeUtils.notifyWebJson(webView, BridgeUtils.WebEvents.ON_QR_DETECTED, payload)
        Log.d(TAG, "QR detected (new): $value")
    }

    // ---- QR detected overlay ----

    /**
     * Called on every ML Kit frame that has a barcode.
     * If overlay is already visible: only reschedule the hide — no layout/color update (prevents flicker).
     * If overlay is not visible: position it, set color, show it.
     * lastBarcodeScreenRect is always updated so zoom repaints use the freshest position.
     */
    private fun showQrOverlay(screenRect: RectF, orange: Boolean) {
        lastBarcodeScreenRect = screenRect

        // Reschedule hide regardless
        overlayHideRunnable?.let { overlayHideHandler.removeCallbacks(it) }
        val hideRunnable = Runnable { hideQrOverlayOnUiThread() }
        overlayHideRunnable = hideRunnable
        overlayHideHandler.postDelayed(hideRunnable, OVERLAY_HIDE_DELAY_MS)

        if (overlayVisible && orange == overlayPaintedOrange) return  // already painted at correct color — skip to prevent flicker

        repaintQrOverlay(screenRect, orange)
    }

    /**
     * Force-reposition and repaint the overlay. Called on first show and on zoom change.
     * screenRect is in PreviewView/RelativeLayout space — no offset adjustment needed.
     */
    private fun repaintQrOverlay(screenRect: RectF, orange: Boolean) {
        val color = Color.parseColor(if (orange) COLOR_QR_DETECTED else COLOR_QR_IN_FRAME)
        overlayVisible = true
        overlayPaintedOrange = orange
        activity.runOnUiThread {
            debugBarcodeOverlay?.let { overlay ->
                val lp = overlay.layoutParams as android.widget.RelativeLayout.LayoutParams
                lp.leftMargin = screenRect.left.toInt()
                lp.topMargin  = screenRect.top.toInt()
                lp.width      = screenRect.width().toInt()
                lp.height     = screenRect.height().toInt()
                overlay.layoutParams = lp
                overlay.setBackgroundColor(color)
                overlay.visibility = View.VISIBLE
            }
        }
    }

    private fun hideQrOverlay() {
        overlayHideRunnable?.let { overlayHideHandler.removeCallbacks(it) }
        overlayHideRunnable = null
        overlayIsOrange = false
        overlayPaintedOrange = false
        lastBarcodeScreenRect = null
        hideQrOverlayOnUiThread()
    }

    private fun hideQrOverlayOnUiThread() {
        overlayVisible = false
        activity.runOnUiThread {
            debugBarcodeOverlay?.visibility = View.GONE
        }
    }

    // ---- Debug overlay helpers ----

    private fun positionDebugOverlay(jsRect: RectF) {
        activity.runOnUiThread {
            debugOverlay?.let { overlay ->
                val lp = overlay.layoutParams as android.widget.RelativeLayout.LayoutParams
                lp.leftMargin = jsRect.left.toInt()
                lp.topMargin = jsRect.top.toInt()
                lp.width = jsRect.width().toInt()
                lp.height = jsRect.height().toInt()
                overlay.layoutParams = lp
                overlay.setBackgroundResource(R.drawable.debug_viewfinder_border)
                overlay.visibility = View.VISIBLE
            }
        }
    }

    private fun updateDebugQrStatus(inside: Boolean) {
        activity.runOnUiThread {
            if (inside) {
                debugQrStatus?.setTextColor(Color.parseColor("#34C759"))
                debugQrStatus?.text = "⬤ QR inside"
            } else {
                debugQrStatus?.setTextColor(Color.parseColor("#FF3B30"))
                debugQrStatus?.text = "⬤ QR outside"
            }
        }
    }
}

File: src/native/iosnativeWebView/Sources/Core/Camera/NativeCameraManager.swift
import Foundation
import AVFoundation
import UIKit
import os

private let logger = Logger(subsystem: Bundle.main.bundleIdentifier ?? "com.app", category: "NativeCameraManager")

/// Mutable reference wrapper so all sub-controller closures share the same callback
/// even after rewireEvents()/rewireError() replace it post-construction.
/// NSLock guards concurrent handler replacement vs. invocation.
private final class EventBox {
    private let lock = NSLock()
    private var _handler: (String, [String: Any]?) -> Void
    init(_ h: @escaping (String, [String: Any]?) -> Void) { _handler = h }

    func call(_ event: String, _ payload: [String: Any]?) {
        lock.lock()
        let h = _handler
        lock.unlock()
        h(event, payload)
    }

    func replace(_ h: @escaping (String, [String: Any]?) -> Void) {
        lock.lock()
        _handler = h
        lock.unlock()
    }
}
private final class ErrorBox {
    private let lock = NSLock()
    private var _handler: (String) -> Void
    init(_ h: @escaping (String) -> Void) { _handler = h }

    func call(_ msg: String) {
        lock.lock()
        let h = _handler
        lock.unlock()
        h(msg)
    }

    func replace(_ h: @escaping (String) -> Void) {
        lock.lock()
        _handler = h
        lock.unlock()
    }
}

/// Thin facade — wires all camera sub-components and exposes the public API
/// that NativeBridge.swift calls. Mirrors Android's NativeCameraManager.kt.
public class NativeCameraManager {

    private let stateMachine: VideoStreamStateMachine
    private let zoomController: ZoomController
    private let torchController: TorchController
    private let barcodeDetector: BarcodeDetector
    private let sessionManager: CameraSessionManager
    private let holdController: HoldController

    // Boxes — all sub-controller closures close over these references,
    // so updating .handler here propagates to all of them immediately.
    private let eventBox: EventBox
    private let errorBox: ErrorBox

    public init(onEvent: @escaping (String, [String: Any]?) -> Void,
                onError: @escaping (String) -> Void) {
        let eBox = EventBox(onEvent)
        let errBox = ErrorBox(onError)
        self.eventBox = eBox
        self.errorBox = errBox

        let sm = VideoStreamStateMachine()
        self.stateMachine = sm

        zoomController = ZoomController(
            stateMachine: sm,
            onZoomChanged: { zoomLevel, minZoom, maxZoom in
                let payload: [String: Any] = [
                    "zoomLevel": round(Double(zoomLevel) * 10) / 10,
                    "minZoom":   round(Double(minZoom)   * 10) / 10,
                    "maxZoom":   round(Double(maxZoom)   * 10) / 10
                ]
                eBox.call("ON_ZOOM_CHANGED", payload)
            }
        )

        torchController = TorchController(
            onTorchChanged: { enabled in
                eBox.call("ON_TORCH_CHANGED", ["enabled": enabled])
            }
        )

        barcodeDetector = BarcodeDetector()

        sessionManager = CameraSessionManager(
            stateMachine: sm,
            zoomController: zoomController,
            torchController: torchController,
            barcodeDetector: barcodeDetector,
            onReady:   { eBox.call("ON_VIDEO_STREAM_READY",   nil) },
            onStopped: { eBox.call("ON_VIDEO_STREAM_STOPPED", nil) },
            onError:   { msg in errBox.call(msg) }
        )

        holdController = HoldController(stateMachine: sm, barcodeDetector: barcodeDetector)

        // Wire detection handler now that holdController exists
        barcodeDetector.detectionHandler = { [weak holdController, weak sm] value, type, bounds in
            guard sm?.state != .hold else { return }
            holdController.map { hc in
                if !hc.isNewValue(value) { hc.startHold(); return }
                hc.startHold()
                let payload: [String: Any] = ["value": value, "format": BarcodeDetector.formatName(type)]
                eBox.call("ON_QR_DETECTED", payload)
                logger.debug("QR detected (new): \(value)")
            }
        }
    }

    // MARK: - Public API

    public func setPreviewView(_ view: UIView) {
        sessionManager.setPreviewView(view)
    }

    public func start(facing: String = "back",
                      autoZoom: Bool = false,         // accepted, silently ignored on iOS
                      initialZoom: Float = 1.0,
                      scanFormat: String = "all",
                      fpsMin: Int? = nil,
                      fpsMax: Int? = nil) {
        guard !stateMachine.isActive else { return }
        stateMachine.transition(to: .starting)
        sessionManager.start(
            facing: facing,
            initialZoom: initialZoom,
            scanFormat: scanFormat,
            fpsMin: fpsMin,
            fpsMax: fpsMax
        )
    }

    public func stop() {
        guard stateMachine.isActive else { return }
        holdController.reset()
        stateMachine.transition(to: .stopping)
        sessionManager.stop()
    }

    public func flip() {
        guard stateMachine.isActive else { return }
        holdController.reset()
        stateMachine.transition(to: .flipping)
        sessionManager.flip()
    }

    public func setZoom(multiplier: Float) {
        zoomController.setZoom(multiplier: multiplier)
    }

    public func setTorch(_ on: Bool) {
        torchController.setTorch(on)
    }

    public func setFps(min: Int?, max: Int?) {
        guard stateMachine.isActive else { return }
        sessionManager.setFps(min: min, max: max)
    }

    public func handlePinchBegan() {
        zoomController.handlePinchBegan()
    }

    public func handlePinchChanged(scale: CGFloat) {
        zoomController.handlePinchChanged(scale: scale)
    }

    public func cleanup() {
        sessionManager.cleanup()
    }

    /// Replace the event callback post-construction (called by NativeBridge after init).
    /// Updates the shared EventBox so all sub-controller closures see the new handler.
    public func rewireEvents(_ handler: @escaping (String, [String: Any]?) -> Void) {
        eventBox.replace(handler)
    }

    /// Replace the error callback post-construction (called by NativeBridge after init).
    /// Updates the shared ErrorBox so all sub-controller closures see the new handler.
    public func rewireError(_ handler: @escaping (String) -> Void) {
        errorBox.replace(handler)
    }

    public var isStreaming: Bool { stateMachine.isActive }
}
